lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs

```diff
@@ -287,24 +400,382 @@ withNetworkLayer tr network np project k =
+    delegations <- concurrently (BF.getTxDelegations hash) (BF.getTxStakes hash)
+    certs <- liftEither $ for (uncurry align delegations) \case
+        This txDelegation -> pure <$> parseTxDelegation txDelegation
+        That txStake -> pure <$> parseTxStake txStake
+        These txDelegation txStake ->
+            (\d s -> [d, s])
+                <$> parseTxDelegation txDelegation
+                <*> parseTxStake txStake
+    pure $ snd <$> sortOn fst (concat certs)
+  where
+    parseTxDelegation BF.TransactionDelegation{..} = do
+        let addr = BF.unAddress _transactionDelegationAddress
+        rewardAccount <-
+            first (InvalidAddress addr) $ decodeStakeAddress @nd addr
+        poolId <- fromBlockfrostM _transactionDelegationPoolId
+        pure ( _transactionDelegationCertIndex
+             , CertDelegateFull rewardAccount poolId
+             )
+    parseTxStake BF.TransactionStake{..} = do
+        let addr = BF.unAddress _transactionStakeAddress
+        rewardAccount <-
+            first (InvalidAddress addr) $ decodeStakeAddress @nd addr
+        let action = if _transactionStakeRegistration
+                then CertRegisterKey
+                else CertDelegateNone
+        pure (_transactionStakeCertIndex, action rewardAccount)
+
+fetchTransaction
+    :: forall m
+     . (MonadError BlockfrostError m, BF.MonadBlockfrost m)
+    => SomeNetworkDiscriminant
+    -> BF.TxHash
+    -> m (BF.Transaction, Tx)
+fetchTransaction nd hash = do
+  transaction <- BF.getTx hash
+  utxos <- BF.getTxUtxos hash
+  withdrawals <- BF.getTxWithdrawals hash
+  metadata <- BF.getTxMetadataJSON hash
+  (transaction,) <$>
+    assembleTransaction nd transaction utxos withdrawals metadata
+
+assembleTransaction
+    :: forall m
+     . MonadError BlockfrostError m
+    => SomeNetworkDiscriminant
+    -> BF.Transaction
+    -> BF.TransactionUtxos
+    -> [BF.TransactionWithdrawal]
+    -> [BF.TransactionMetaJSON]
+    -> m Tx
+assembleTransaction
+    (SomeNetworkDiscriminant (Proxy :: Proxy nd))
+    BF.Transaction{..}
+    BF.TransactionUtxos{..}
+    txWithdrawals
+    metadataJSON = liftEither do
+        txId <- parseTxHash _transactionHash
+        let fee = Just $ Coin $ fromIntegral _transactionFees
+        (resolvedInputs, resolvedCollateralInputs) <-
+            fromInputs _transactionUtxosInputs
+        outputs <- for _transactionUtxosOutputs \out@BF.UtxoOutput{..} -> do
+            let outAddr = BF.unAddress _utxoOutputAddress
+            address <- either (throwError . InvalidAddress outAddr) pure $
+                decodeAddress @nd outAddr
+            tokens <- do
+                coin <- case [ lovelaces
+                             | BF.AdaAmount lovelaces <- _utxoOutputAmount ] of
+                    [l] -> fromBlockfrost l
+                    _ -> throwError $ InvalidUtxoOutputAmount out
+                pure $ TokenBundle coin mempty -- TODO: Handle native assets
```

Comment:
Could you make a ticket for the TODO?

Reply:
https://input-output.atlassian.net/browse/ADP-1777